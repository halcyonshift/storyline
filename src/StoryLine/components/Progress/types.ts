import { SectionModel } from '@sl/db/models'

export type ProgressProps = {
    section: SectionModel
}
