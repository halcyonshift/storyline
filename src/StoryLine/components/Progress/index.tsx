import { Box, Tooltip, Typography } from '@mui/material'
import { useTranslation } from 'react-i18next'
import useSettings from '@sl/theme/useSettings'
import { status } from '@sl/theme/utils'
import { ProgressProps } from './types'

const Progress = ({ section }: ProgressProps) => {
    const settings = useSettings()
    const { t } = useTranslation()

    const getPercentage = () => {
        try {
            return (100 / section.wordGoal) * section.wordCount > 100
                ? 100
                : (100 / section.wordGoal) * section.wordCount
        } catch {
            return 0
        }
    }

    const getBackgroundColor = () => {
        if (section.isPart) return status(section.status, settings.isDark() ? 900 : 200).color
        else if (section.isChapter)
            return status(section.status, settings.isDark() ? 800 : 100).color
        return status(section.status, settings.isDark() ? 600 : 50).color
    }

    const getTooltipTitle = () => {
        if (section.daysRemaining && !section.wordGoal) {
            return `${section.wordCount.toLocaleString()}: ${t(
                'view.work.dashboard.tracker.daysRemaining',
                {
                    days: section.daysRemaining.toLocaleString()
                }
            )}`
        }

        const base = `${section.wordCount.toLocaleString()}/${section.wordGoal.toLocaleString()}`

        if (section.wordGoal && section.daysRemaining)
            return `${base}: ${t('view.work.dashboard.tracker.daysWordsRemaining', {
                days: section.daysRemaining.toLocaleString(),
                words: section.wordsPerDay.toLocaleString()
            })}`

        return base
    }

    if (section.wordGoal) {
        return (
            <Tooltip title={getTooltipTitle()}>
                <Box className='bg-slate-100 dark:bg-slate-800 relative'>
                    <Box
                        className='absolute w-full h-full border-r-2 border-white dark:border-black'
                        sx={{
                            backgroundColor: getBackgroundColor(),
                            width: `${getPercentage()}%`
                        }}></Box>
                    <Box className='flex justify-between px-2 py-1 relative z-10'>
                        <Typography
                            variant='body1'
                            className='whitespace-nowrap overflow-hidden text-ellipsis'>
                            {section.displayTitle}
                        </Typography>
                        <Typography variant='body1' className='whitespace-nowrap'>
                            {section.wordCount ? section.wordCount.toLocaleString() : 0}
                        </Typography>
                    </Box>
                </Box>
            </Tooltip>
        )
    }

    if (section.daysRemaining) {
        return (
            <Tooltip title={getTooltipTitle()}>
                <Box
                    className='flex justify-between px-2 py-1'
                    sx={{ backgroundColor: getBackgroundColor() }}>
                    <Typography
                        variant='body1'
                        className='whitespace-nowrap overflow-hidden text-ellipsis'>
                        {section.displayTitle}
                    </Typography>
                    <Typography variant='body1' className='whitespace-nowrap'>
                        {section.wordCount ? section.wordCount.toLocaleString() : 0}
                    </Typography>
                </Box>
            </Tooltip>
        )
    }

    return (
        <Box
            className='flex justify-between px-2 py-1'
            sx={{ backgroundColor: getBackgroundColor() }}>
            <Typography variant='body1' className='whitespace-nowrap overflow-hidden text-ellipsis'>
                {section.displayTitle}
            </Typography>
            <Typography variant='body1' className='whitespace-nowrap'>
                {section.wordCount ? section.wordCount.toLocaleString() : 0}
            </Typography>
        </Box>
    )
}

export default Progress
