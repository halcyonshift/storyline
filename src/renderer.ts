import { init } from '@sentry/electron/renderer'
import { init as reactInit } from '@sentry/react'

if (parseInt(process.env.DEBUG) === 0) {
    init({ dsn: process.env.SENTRY }, reactInit)
}

import '@fontsource/roboto/300.css'
import '@fontsource/roboto/400.css'
import '@fontsource/roboto/500.css'
import '@fontsource/roboto/700.css'
import './index.css'
import './StoryLine'
