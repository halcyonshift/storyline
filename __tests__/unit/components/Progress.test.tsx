import renderer from 'react-test-renderer'
import Progress from '@sl/components/Progress'
import { render, screen } from '../test-utils'

describe('<Progress />', () => {
    it('is green if percentage 100%', () => {
        render(<Progress words={100} goal={100} />)
        expect(screen.getByRole('progressbar')).toHaveClass('MuiLinearProgress-colorSuccess')
    })

    it('is blue if percentage below 100%', () => {
        render(<Progress words={75} goal={100} />)
        expect(screen.getByRole('progressbar')).toHaveClass('MuiLinearProgress-colorInfo')
    })

    it('is orange if percentage below 75%', () => {
        render(<Progress words={50} goal={100} />)
        expect(screen.getByRole('progressbar')).toHaveClass('MuiLinearProgress-colorWarning')
    })

    it('is orange if percentage below 50%', () => {
        render(<Progress words={49} goal={100} />)
        expect(screen.getByRole('progressbar')).toHaveClass('MuiLinearProgress-colorError')
    })

    it("doesn't display if no goal ", () => {
        render(<Progress words={100} goal={0} />)
        expect(screen.queryByRole('progressbar')).toBeFalsy()
    })

    it('renders correctly', () => {
        const tree = renderer.create(<Progress words={100} goal={1000} />).toJSON()
        expect(tree).toMatchSnapshot()
    })
})
