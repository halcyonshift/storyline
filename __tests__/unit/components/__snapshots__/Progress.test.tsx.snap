// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Progress /> renders correctly 1`] = `
<div
  className="MuiBox-root css-1gigelk"
>
  <div
    aria-label="100/1,000 (10.0%)"
    aria-labelledby={null}
    className="w-full MuiBox-root css-0"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <span
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={10}
      className="MuiLinearProgress-root MuiLinearProgress-colorError MuiLinearProgress-determinate css-1qzfjjf-MuiLinearProgress-root"
      role="progressbar"
    >
      <span
        className="MuiLinearProgress-bar MuiLinearProgress-barColorError MuiLinearProgress-bar1Determinate css-ete1bz-MuiLinearProgress-bar1"
        style={
          {
            "transform": "translateX(-90%)",
          }
        }
      />
    </span>
  </div>
</div>
`;
